From 72f6b05f1baef8003ec689c06be1cbeb9f673fdc Mon Sep 17 00:00:00 2001
From: Kshitij Gupta <kshitijgm@gmail.com>
Date: Wed, 15 Aug 2018 11:47:05 +0000
Subject: [PATCH] Gralloc2: Allow invalid usage bits

08-15 16:59:23.135   325   353 E Gralloc2: buffer descriptor contains invalid usage bits 0x2000000
08-15 16:59:23.135   325   353 E GraphicBufferAllocator: Failed to allocate (1920 x 1080) layerCount 1 format 2141391876 usage 42002900: 3
08-15 16:59:23.135   325   353 E BufferQueueProducer: [SurfaceView - com.google.android.gallery3d/com.android.gallery3d.app.MovieActivity#0] dequeueBuffer: createGraphicBuffer failed
08-15 16:59:23.135   425  9850 E ACodec  : dequeueBuffer failed: NO_MEMORY(-12).

- Many older grallocs seem to be providing invalid usage bits, thus
  breaking video playback/decoding and similar functionality. Don't return
  errors if we come across invalid bits. Log a warning and continue.

Test: Boot and play a video from storage
Change-Id: Iaffbd3b4709c34dd1809e547a1c35e64b2529636
Signed-off-by: Kshitij Gupta <kshitijgm@gmail.com>
---
 libs/ui/Gralloc2.cpp | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/libs/ui/Gralloc2.cpp b/libs/ui/Gralloc2.cpp
index b92cbf351..d577941bb 100644
--- a/libs/ui/Gralloc2.cpp
+++ b/libs/ui/Gralloc2.cpp
@@ -90,9 +90,8 @@ Gralloc2::Error Mapper::validateBufferDescriptorInfo(
     }
 
     if (descriptorInfo.usage & ~validUsageBits) {
-        ALOGE("buffer descriptor contains invalid usage bits 0x%" PRIx64,
+        ALOGW("buffer descriptor contains invalid usage bits 0x%" PRIx64,
               descriptorInfo.usage & ~validUsageBits);
-        return Error::BAD_VALUE;
     }
     return Error::NONE;
 }
-- 
2.17.1

